### DISCUSSION TITLE
doc:key shared

### DISCUSSION THREAD

**Discussion by startjava** [apache/pulsar#19070] [General]
the broker will allow it to receive messages knowing some messages of that key may be still be processing in other consumers at the time, thus order may be affected for that short period of adding a new consumer.

" the broker will allow it to receive messages"  allow new consumer receive read position after message ??
" the broker will allow it to receive messages"  allow new consumer receive same key or no same key message ??

@asafm 
i have top 2 question , thank you !

**Comment by asafm** [comment]
For clarity for other people as well, I'll attach the full quote taken from the [documentation](https://pulsar.apache.org/docs/next/concepts-messaging/#preserving-order-of-processing) site regarding Key Shared Subscription

> That requirement can be relaxed by enabling `allowOutOfOrderDelivery` via the Consumer API. If set on the new consumer, then when it is connected, the broker will allow it to receive messages knowing some messages of that key may be still be processing in other consumers at the time, thus order may be affected for that short period of adding a new consumer.

Answering your first question: 
> " the broker will allow it to receive messages" allow new consumer receive read position after message ??

I'm assuming you mean to ask if the new consumer will be allowed to receive messages after the read position when that flag is enabled; then, from my understanding, yes.

Regarding your second question:
> " the broker will allow it to receive messages" allow new consumer receive same key or no same key message ??

The new consumer will be in charge of certain keys, right? Those keys are always taken from existing consumers. So it might be possible that certain messages of key K1 were delivered to C2, and when C_New joined, messages of K1 from now on will be delivered to C_New. The thing is that when that flag is enabled, the broker won't wait for C2 to finish processing messages with key K1 (acknowledge) but will simply start delivering messages from read position to C_New with key K1. 
 
I hope I managed to answer, if not please help me by clarifying your questions @startjava .

**Reply by startjava** [reply]
no time-out was defined (acknowledging for you)

"acknowledging for you" mean "consumer ack"?

**Reply by asafm** [reply]
>no time-out was defined (acknowledging for you)

Yes, there is no timeout. That's the trade-off. 

>"acknowledging for you" mean "consumer ack"?

Yes, consumer ack.

**Comment by startjava** [comment]
i understand：
（1）disable allowOutOfOrderDelivery：all message ack , next step ,new consumer receive message
（2）enable allowOutOfOrderDelivery：no all message ack, next step ,new consumer receive message
Am I getting this right?thank you !

**Reply by asafm** [reply]
Yes. You can try that pretty easily :)
